### PR TITLE
Improve PM UI Experience with Package searches

### PR DESCRIPTION
**Pull request by sbanni**
## Bug

Fixes: https://github.com/nuget/home/issues/10008
Regression: No  

## Fix

Details: Instead of waiting for the UI to ask for the data to display and kicking off the work in the background and popping in when ready we instead kick the work off right away and cache the results into an in memory cache so when requested it is already readily available.

In addition, we also retrieve and cache the detailed package metadata for the version that will be display first when selecting the package item, this greatly improves the experience for packages with large # of versions in the first click scenario and in re-selecting the package.  Before we would re-fetch the data if you selected a different package and then re-selected it.
Before:
slow
After:
faster

## Testing/Validation

Tests Added: No
Reason for not adding tests: There are tests that already cover this area  
Validation:  Manual testing and unit tests

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs [hunk 1]
```diff
@@ -51,14 +53,32 @@ protected DetailControlModel(IEnumerable<IProjectContextInfo> projects)
...
         /// The method is called when the associated DocumentWindow is closed.
         /// </summary>
         public virtual void CleanUp()
         {
-            Options.SelectedChanged -= DependencyBehavior_SelectedChanged;
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs]: `disposing: true`

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs [hunk 2]
```diff
@@ -191,11 +214,29 @@ public virtual void CleanUp()
                 .Select(GetVersion)
                 .ToList();
 
-            // hook event handler for dependency behavior changed
-            Options.SelectedChanged += DependencyBehavior_SelectedChanged;
-
             await CreateVersionsAsync(CancellationToken.None);
             OnCurrentPackageChanged();
+
+            var detailedPackageSearchMetadata = await searchResultPackage.GetDetailedPackageSearchMetadataAsync();
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs]: Use the explicit type and declare and use better member names than `Item1` and `Item2`.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs [hunk 3]
```diff
@@ -191,11 +214,29 @@ public virtual void CleanUp()
...
+            if (detailedPackageSearchMetadata.Item1 != null)
+            {
+                var deprecationMetadata = detailedPackageSearchMetadata.Item2;
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs]: Use explicit type.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs [hunk 4]
```diff
@@ -1028,20 +1029,26 @@ private void SettingsButtonClicked(object sender, EventArgs e)
...
         private void PackageList_SelectionChanged(object sender, SelectionChangedEventArgs e)
         {
+            var loadCts = new CancellationTokenSource();
+            var oldCts = Interlocked.Exchange(ref _cancelSelectionChangedSource, loadCts);
+            oldCts?.Cancel();
+            oldCts?.Dispose();
+
             NuGetUIThreadHelper.JoinableTaskFactory
-                .RunAsync(UpdateDetailPaneAsync)
+                .RunAsync(async () => await UpdateDetailPaneAsync(_packageList.SelectedItem, loadCts.Token))
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs]: Why pass `_packageList.SelectedItem` as an argument?  Why not just access it in `UpdateDetailPaneAsync(...)` [like before](https://github.com/NuGet/NuGet.Client/pull/3729/files#diff-80991ce1edef4f242608751b70dd2d48dad8e24a28dbefd2b68a3a7784a12f21L1041)?  You're [already accessing `_packageList.SelectedIndex` there](https://github.com/NuGet/NuGet.Client/pull/3729/files#diff-80991ce1edef4f242608751b70dd2d48dad8e24a28dbefd2b68a3a7784a12f21R1048).

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs [hunk 5]
```diff
@@ -1050,15 +1057,17 @@ internal async Task UpdateDetailPaneAsync()
...
-                await _detailModel.SetCurrentPackage(selectedPackage, _topPanel.Filter, () => _packageList.SelectedItem);
-                _detailModel.SetCurrentSelectionInfo(selectedIndex, recommendedCount, _recommendPackages, selectedPackage.RecommenderVersion);
+                await _detailModel.SetCurrentPackage(selectedItem, _topPanel.Filter, () => _packageList.SelectedItem);
+                _detailModel.SetCurrentSelectionInfo(selectedIndex, recommendedCount, _recommendPackages, selectedItem.RecommenderVersion);
 
                 _packageDetail.ScrollToHome();
 
-                var context = new PackageLoadContext(Model.IsSolution, Model.Context);
-                await _detailModel.LoadPackageMetadataAsync(SelectedSource.PackageSources, CancellationToken.None);
+                if(cancellationToken.IsCancellationRequested)
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs]: Format document.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs [hunk 6]
```diff
@@ -1050,15 +1057,17 @@ internal async Task UpdateDetailPaneAsync()
                 _packageDetail.Visibility = Visibility.Visible;
                 _packageDetail.DataContext = _detailModel;
 
-                EmitSearchSelectionTelemetry(selectedPackage);
+                EmitSearchSelectionTelemetry(selectedItem);
 
-                await _detailModel.SetCurrentPackage(selectedPackage, _topPanel.Filter, () => _packageList.SelectedItem);
-                _detailModel.SetCurrentSelectionInfo(selectedIndex, recommendedCount, _recommendPackages, selectedPackage.RecommenderVersion);
+                await _detailModel.SetCurrentPackage(selectedItem, _topPanel.Filter, () => _packageList.SelectedItem);
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs]: Rename to `SetCurrentPackageAsync(...)`.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs [hunk 7]
```diff
@@ -43,6 +43,7 @@ public partial class PackageManagerControl : UserControl, IVsWindowSearch
         internal event EventHandler _actionCompleted;
         internal DetailControlModel _detailModel;
         internal CancellationTokenSource _loadCts;
+        private CancellationTokenSource _cancelSelectionChangedSource;
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:46]: When this control closes, be sure to dispose of this.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageSearchService.cs [hunk 8]
```diff
@@ -31,6 +33,13 @@ public sealed class NuGetPackageSearchService : INuGetSearchService
         private readonly ISharedServiceState _sharedServiceState;
         private readonly Microsoft.VisualStudio.Threading.AsyncLazy<SourceRepository> _packagesFolderLocalRepositoryLazy;
         private readonly Microsoft.VisualStudio.Threading.AsyncLazy<IReadOnlyList<SourceRepository>> _globalPackageFolderRepositoriesLazy;
+        private readonly static MemoryCache PackageSearchMetadataMemoryCache = new MemoryCache("PackageSearchMetadata",
+             new NameValueCollection
+             {
+                {"cacheMemoryLimitMegabytes", "4"},
+                {"physicalMemoryLimitPercentage", "0"},
+                {"pollingInterval", "00:02:00"}
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageSearchService.cs]: One level of indentation is 4 spaces.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageSearchService.cs [hunk 9]
```diff
@@ -69,10 +78,53 @@ public NuGetPackageSearchService(ServiceActivationOptions options, IServiceBroke
...
+            CancellationToken cancellationToken)
+        {
+            Assumes.NotNull(identity);
+            Assumes.NotNullOrEmpty(packageSources);
+
+            string cacheId = PackageSearchMetadataCacheObject.GetCacheId(identity.Id, includePrerelease, packageSources);
+            if (PackageSearchMetadataMemoryCache.Get(cacheId) is PackageSearchMetadataCacheObject backgroundDataCache)
+            {
+                var packageSearchData = await backgroundDataCache.DetailedPackageSearchMetadataContextInfo;
+                var deprecatedData = await backgroundDataCache.PackageDeprecrationMetadataContextInfo;
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageSearchService.cs]: Use explicit types.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageItemListViewModel.cs [hunk 10]
```diff
@@ -462,7 +464,12 @@ public void UpdatePackageStatus(IEnumerable<PackageCollectionItem> installedPack
                     return latestAvailableVersion;
                 });
 
-            _backgroundDeprecationMetadataLoader = AsyncLazy.New(GetPackageDeprecationMetadataAsync);
+            _backgroundDeprecationMetadataLoader = AsyncLazy.New(
+                async () =>
+                {
+                    var packageSearchMetadata = await GetDetailedPackageSearchMetadataAsync();
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageItemListViewModel.cs]: Use explicit type.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs [hunk 11]
```diff
@@ -478,22 +519,71 @@ public DisplayVersion SelectedVersion
...
                     }
-                    else
+                }
+            }
+        }
+
+        private async ValueTask SelectedVersionChangedAsync(PackageItemListViewModel packageItemListViewModel, NuGetVersion nugetVersion, CancellationToken cancellationToken)
+        {
+            // Load the detailed metadata that we already have and check to see if this matches what is selected, we cannot use the _metadataDict here unfortunately as it won't be populated yet
+            (PackageSearchMetadataContextInfo, PackageDeprecationMetadataContextInfo) detailedPackageSearchMetadata = await packageItemListViewModel.GetDetailedPackageSearchMetadataAsync();
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs]: Declare and use better member names than `Item1` and `Item2`.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs [hunk 12]
```diff
@@ -478,22 +519,71 @@ public DisplayVersion SelectedVersion
...
+
+                PackageMetadata = new DetailedPackageMetadata(
+                    detailedPackageSearchMetadata.Item1,
+                    detailedPackageSearchMetadata.Item2,
+                    packageItemListViewModel.DownloadCount);
+            }
+            else
+            {
+                // We don't have the data readily available, we need to query the server
+                IServiceBroker serviceBroker = await BrokeredServicesUtilities.GetRemoteServiceBrokerAsync();
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs:563]: The feature branch does not yet contain https://github.com/NuGet/NuGet.Client/commit/8ab36fe54a0f42e87b2150c3d2d9b1d61f643c1b; however, we should not lose sight of this.  We should try to use the cached instance.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs [hunk 13]
```diff
@@ -478,22 +519,71 @@ public DisplayVersion SelectedVersion
...
+                using (INuGetSearchService searchService = await serviceBroker.GetProxyAsync<INuGetSearchService>(NuGetServices.SearchService, cancellationToken))
+                {
+                    PackageIdentity id = new PackageIdentity(packageItemListViewModel.Id, nugetVersion);
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs]: `var`

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs [hunk 14]
```diff
@@ -478,22 +519,71 @@ public DisplayVersion SelectedVersion
...
+                        var oldCts = Interlocked.Exchange(ref _selectedVersionCancellationTokenSource, loadCts);
+                        oldCts?.Cancel();
+                        oldCts?.Dispose();
+
+                        if (_metadataDict.TryGetValue(_selectedVersion.Version, out DetailedPackageMetadata detailedPackageMetadata))
+                        {
+                            PackageMetadata = detailedPackageMetadata;
+                        }
+
+                        NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(() => SelectedVersionChangedAsync(_searchResultPackage, _selectedVersion.Version, loadCts.Token).AsTask());
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs:537]: Use that `PostOnFailure` extension method.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs [hunk 15]
```diff
@@ -478,22 +519,71 @@ public DisplayVersion SelectedVersion
...
+
+                    if (cancellationToken.IsCancellationRequested || _searchResultPackage != packageItemListViewModel)
                     {
-                        PackageMetadata = null;
+                        return;
                     }
 
-                    OnPropertyChanged(nameof(SelectedVersion));
-                    OnSelectedVersionChanged();
+                    DetailedPackageMetadata detailedPackageMetadata = new DetailedPackageMetadata(
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs]: `var`

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs [hunk 16]
```diff
@@ -18,18 +19,37 @@
...
+
+        public SearchObject(
+            IPackageFeed mainFeed,
+            IPackageFeed? recommenderFeed,
+            IPackageMetadataProvider packageMetadataProvider,
+            IReadOnlyCollection<PackageSourceContextInfo> packageSources,
+            MemoryCache? searchCache)
         {
             Assumes.NotNull(mainFeed);
+            Assumes.NotNullOrEmpty(packageSources);
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs:47]: Add `Assumes.NotNull(packageMetadataProvider);`

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs [hunk 17]
```diff
@@ -45,16 +65,12 @@ public async ValueTask<SearchResultContextInfo> SearchAsync(string searchText, S
...
             if (recommenderFeedResults != null)
             {
                 // remove duplicated recommended packages from the browse results
-                var recommendedIds = recommenderFeedResults.Items.Select(item => item.Identity.Id);
-
-                IReadOnlyCollection<PackageSearchMetadataContextInfo> filteredMainFeedResult = mainFeedResult.Items
-                    .Where(item => !recommendedIds.Contains(item.Identity.Id))
-                    .Select(mainFeedPackageSearchMetadata => PackageSearchMetadataContextInfo.Create(mainFeedPackageSearchMetadata))
-                    .ToList();
+                IEnumerable<string> recommendedIds = recommenderFeedResults.Items.Select(item => item.Identity.Id).ToList();
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs]: `List<string>`?  The upcast seems useless.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs [hunk 18]
```diff
@@ -66,7 +82,21 @@ public async ValueTask<SearchResultContextInfo> SearchAsync(string searchText, S
                     })
                     .ToList();
 
-                recommendedPackageSearchMetadataContextInfo.AddRange(filteredMainFeedResult);
+                List<IPackageSearchMetadata> filteredMainFeedResults = mainFeedResult.Items.Where(item => !recommendedIds.Contains(item.Identity.Id)).ToList();
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs]: Should `recommendedIds` be a hash set?  Should it be case-insensitive?

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs [hunk 19]
```diff
@@ -66,7 +82,21 @@ public async ValueTask<SearchResultContextInfo> SearchAsync(string searchText, S
...
+
+                recommendedPackageSearchMetadataContextInfo.AddRange(
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs]: How are you calling `AddRange(...)` on [`IList<PackageSearchMetadataContextInfo>`](https://github.com/NuGet/NuGet.Client/pull/3729/files#diff-6d92112d740f940faa169328fd0037e1386a792a091eaf5c5722c7fdeecd473cR75).  `IList<T>` only defines [`Add(...)`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.icollection-1.add?view=netframework-4.7.2#System_Collections_Generic_ICollection_1_Add__0_).

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs [hunk 20]
```diff
@@ -66,7 +82,21 @@ public async ValueTask<SearchResultContextInfo> SearchAsync(string searchText, S
...
+                    filteredMainFeedResults
+                    .Select(mainFeedPackageSearchMetadata => PackageSearchMetadataContextInfo.Create(mainFeedPackageSearchMetadata))
+                    .ToList());
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs]: Is `.ToList()` necessary/beneficial?

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs [hunk 21]
```diff
@@ -66,7 +82,21 @@ public async ValueTask<SearchResultContextInfo> SearchAsync(string searchText, S
...
+
+                List<IPackageSearchMetadata> packageSearchMetadataList = new List<IPackageSearchMetadata>(recommenderFeedResults.Items);
+                packageSearchMetadataList.AddRange(filteredMainFeedResults);
+
+                foreach (IPackageSearchMetadata packageSearchMetadata in packageSearchMetadataList)
+                {
+                    CacheBackgroundDataAsync(packageSearchMetadata, filter.IncludePrerelease);
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs]: -`Async()`?

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs [hunk 22]
```diff
@@ -88,6 +123,7 @@ public async ValueTask<SearchResultContextInfo> SearchAsync(string searchText, S
         public async ValueTask<SearchResultContextInfo> RefreshSearchAsync(CancellationToken cancellationToken)
         {
             Assumes.NotNull(_lastMainFeedSearchResult);
+            Assumes.NotNull(_lastSearchFilter);
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs:120]: Dumb question:  if one refreshes the Browse tab with default search filter, which is no filter, will this code path be hit?  If yes, will `_lastSearchFilter` be `null`?

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Services/NuGetPackageSearchServiceTests.cs [hunk 23]
```diff
@@ -155,7 +162,7 @@ public async Task GetTotalCountAsync_WithLessThanMaxCountResults_ReturnsResultsC
                     ItemFilter.All,
                     CancellationToken.None);
 
-                Assert.Equal(22, totalCount);
+                Assert.Equal(1, totalCount);
```
Comment (dtivel) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Services/NuGetPackageSearchServiceTests.cs:165]: Can you elaborate on this change?  I don't understand it.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageSearchMetadataCacheObject.cs [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+
+        public Task<IReadOnlyCollection<VersionInfoContextInfo>> AllVersionsContextInfo { get; }
+        public Task<PackageDeprecationMetadataContextInfo?> PackageDeprecrationMetadataContextInfo { get; }
+        public Task<PackageSearchMetadataContextInfo> DetailedPackageSearchMetadataContextInfo { get; }
+
+        public static string GetCacheId(string packageId, bool includePrerelease, IReadOnlyCollection<PackageSourceContextInfo> packageSources)
+        {
+            string packageSourcesString = string.Join(" ", packageSources.Select(ps => ps.Name));
+            return string.Concat(packageId, ":", includePrerelease, " - ", packageSourcesString);
+        }
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageSearchMetadataCacheObject.cs:43]: I'm not crazy about routinely making bigger strings out of littler strings as an implementation detail.  We already have UI delay / GC pause bugs from doing it elsewhere.

Is it reasonable to use hash codes instead? `HashCodeCombiner` might work.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageSearchMetadataCacheObject.cs [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+
+        private async Task<IReadOnlyCollection<VersionInfoContextInfo>> GetVersionInfoContextInfoAsync()
+        {
+            IEnumerable<VersionInfo> versions = await _packageSearchMetadata.GetVersionsAsync();
+            IEnumerable<Task<VersionInfoContextInfo>> versionContextInfoTasks = versions.Select(async v => await VersionInfoContextInfo.CreateAsync(v));
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageSearchMetadataCacheObject.cs:48]: Can simplify by eliminating the async lambda?
```C#
IEnumerable<Task<VersionInfoContextInfo>> versionContextInfoTasks = versions.Select(VersionInfoContextInfo.CreateAsync);
```

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageSearchMetadataCacheObject.cs [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+        {
+            _packageSearchMetadata = packageSearchMetadata;
+            _packageMetadataProvider = packageMetadataProvider;
+            AllVersionsContextInfo = GetVersionInfoContextInfoAsync();
+            PackageDeprecrationMetadataContextInfo = GetPackageDeprecationMetadataContextInfoAsync();
+            DetailedPackageSearchMetadataContextInfo = GetDetailedPackageSearchMetadataContextInfoAsync();
+        }
+
+        public Task<IReadOnlyCollection<VersionInfoContextInfo>> AllVersionsContextInfo { get; }
+        public Task<PackageDeprecationMetadataContextInfo?> PackageDeprecrationMetadataContextInfo { get; }
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageSearchMetadataCacheObject.cs]: `PackageDeprecrationMetadataContextInfo` is misspelled.

`Deprecration` -> `Deprecation`

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageSearchMetadataCacheObject.cs [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+        public PackageSearchMetadataCacheObject(IPackageSearchMetadata packageSearchMetadata, IPackageMetadataProvider packageMetadataProvider)
+        {
+            _packageSearchMetadata = packageSearchMetadata;
+            _packageMetadataProvider = packageMetadataProvider;
+            AllVersionsContextInfo = GetVersionInfoContextInfoAsync();
+            PackageDeprecrationMetadataContextInfo = GetPackageDeprecationMetadataContextInfoAsync();
+            DetailedPackageSearchMetadataContextInfo = GetDetailedPackageSearchMetadataContextInfoAsync();
+        }
+
+        public Task<IReadOnlyCollection<VersionInfoContextInfo>> AllVersionsContextInfo { get; }
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageSearchMetadataCacheObject.cs]: These all must be `Task`/`Task<T>` and cannot be `ValueTask`/`ValueTask<T>`?

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageSearchService.cs [hunk 28]
```diff
@@ -217,63 +328,55 @@ public void Dispose()
                 return packageFeeds;
             }
 
-            using (INuGetProjectManagerService? projectManagerService = await _serviceBroker.GetProxyAsync<INuGetProjectManagerService>(
-                NuGetServices.ProjectManagerService,
-                _options,
-                cancellationToken))
+            IReadOnlyCollection<IPackageReferenceContextInfo> installedPackages = await GetAllInstalledPackagesAsync(projectContextInfos, cancellationToken);
+
+            var installedPackageCollection = PackageCollection.FromPackageReferences(installedPackages);
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageSearchService.cs]: Use explicit type.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs (hunk 14 above)
Comment (sbanni) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs:537]: I purposely avoided that pending the discussion about Cancelled token exceptions

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs (hunk 19 above)
Comment (sbanni) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs]: There is an extension in NuGet.Packaging.CollectionExtensions

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs (hunk 21 above)
Comment (sbanni) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs]: What?

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs (hunk 22 above)
Comment (sbanni) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs:120]: This code path will not be hit, lastMainFeedSearchResult would be null first if this was going to be hit and we have checks in the other places where search filter is passed in to ensure not null.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Services/NuGetPackageSearchServiceTests.cs (hunk 23 above)
Comment (sbanni) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Services/NuGetPackageSearchServiceTests.cs:165]: I changed the testdata, the result data set only contains one item now.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageSearchMetadataCacheObject.cs (hunk 25 above)
Comment (sbanni) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageSearchMetadataCacheObject.cs:48]: Nope, it did not like that ;)

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs (hunk 19 above)
Comment (sbanni) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs]: I changed this and removed most of the linq calls

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageSearchMetadataCacheObject.cs (hunk 24 above)
Comment (sbanni) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageSearchMetadataCacheObject.cs:43]: I was also not a huge fan of this, I will look into HashCodeCombiner.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs (hunk 18 above)
Comment (sbanni) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs]: Per our discussion, mainfeedresults is always <= 25 items, I am not too concerned about making it a hashset at this time, we can revisit if we need.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/GlobalSuppressions.cs [hunk 29]
```diff
@@ -39,7 +39,7 @@
...
+[assembly: SuppressMessage("Build", "CA1062:In externally visible method 'Task DetailControlModel.SetCurrentPackage(PackageItemListViewModel searchResultPackage, ItemFilter filter, Func<PackageItemListViewModel> getPackageItemListViewModel)', validate parameter 'getPackageItemListViewModel' is non-null before using it. If appropriate, throw an ArgumentNullException when the argument is null or add a Code Contract precondition asserting non-null argument.", Justification = "<Pending>", Scope = "member", Target = "~M:NuGet.PackageManagement.UI.DetailControlModel.SetCurrentPackageAsync(NuGet.PackageManagement.UI.PackageItemListViewModel,NuGet.PackageManagement.UI.ItemFilter,System.Func{NuGet.PackageManagement.UI.PackageItemListViewModel})~System.Threading.Tasks.Task")]
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/GlobalSuppressions.cs]: Also update the message from `SetCurrentPackage` to `SetCurrentPackageAsync`.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/GlobalSuppressions.cs [hunk 30]
```diff
@@ -54,7 +54,7 @@
...
+[assembly: SuppressMessage("Build", "CA1062:In externally visible method 'Task PackageDetailControlModel.SetCurrentPackage(PackageItemListViewModel searchResultPackage, ItemFilter filter, Func<PackageItemListViewModel> getPackageItemListViewModel)', validate parameter 'getPackageItemListViewModel' is non-null before using it. If appropriate, throw an ArgumentNullException when the argument is null or add a Code Contract precondition asserting non-null argument.", Justification = "<Pending>", Scope = "member", Target = "~M:NuGet.PackageManagement.UI.PackageDetailControlModel.SetCurrentPackageAsync(NuGet.PackageManagement.UI.PackageItemListViewModel,NuGet.PackageManagement.UI.ItemFilter,System.Func{NuGet.PackageManagement.UI.PackageItemListViewModel})~System.Threading.Tasks.Task")]
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/GlobalSuppressions.cs]: Also update the message from `SetCurrentPackage` to `SetCurrentPackageAsync`.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageSearchMetadataCacheObject.cs [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+        }
+
+        private async ValueTask<PackageDeprecationMetadataContextInfo?> GetPackageDeprecationMetadataContextInfoAsync()
+        {
+            PackageDeprecationMetadata? deprecationMetadata = await _packageSearchMetadata.GetDeprecationMetadataAsync();
+            if (deprecationMetadata == null)
+            {
+                return null;
+            }
+            return NuGet.VisualStudio.Internal.Contracts.PackageDeprecationMetadataContextInfo.Create(deprecationMetadata);
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageSearchMetadataCacheObject.cs:59]: Can simplify type name by removing namespace?

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs [hunk 32]
```diff
@@ -18,65 +19,93 @@
...
                 // remove duplicated recommended packages from the browse results
-                var recommendedIds = recommenderFeedResults.Items.Select(item => item.Identity.Id);
+                List<string> recommendedIds = recommenderFeedResults.Items.Select(item => item.Identity.Id).ToList();
+                var recommendedPackageSearchMetadataContextInfo = new List<PackageSearchMetadataContextInfo>();
 
-                IReadOnlyCollection<PackageSearchMetadataContextInfo> filteredMainFeedResult = mainFeedResult.Items
-                    .Where(item => !recommendedIds.Contains(item.Identity.Id))
-                    .Select(mainFeedPackageSearchMetadata => PackageSearchMetadataContextInfo.Create(mainFeedPackageSearchMetadata))
-                    .ToList();
+                foreach (var recommendedFeedResultItem in recommenderFeedResults.Items)
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs]: Use explicit type.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs [hunk 33]
```diff
@@ -18,65 +19,93 @@
...
+                    recommendedPackageSearchMetadataContextInfo.Add(
+                        PackageSearchMetadataContextInfo.Create(
+                            recommendedFeedResultItem,
+                            isRecommended: true,
+                            recommenderVersion: (_recommenderFeed as RecommenderPackageFeed)?.VersionInfo));
+                }
 
-                IList<PackageSearchMetadataContextInfo> recommendedPackageSearchMetadataContextInfo = recommenderFeedResults.Items
-                    .Select(packageSearchMetadata =>
+                foreach (var mainFeedResultItem in mainFeedResult.Items)
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/SearchObject.cs]: Use explicit type.

**Review by dtivel — APPROVED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageSearchMetadataCacheObject.cs (hunk 31 above)
Comment (sbanni) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageSearchMetadataCacheObject.cs:59]: It complains about the property with the same name, will leave as is.

**Review by rrelyea — APPROVED** [review]
:shipit:

**Pull request merged**